Add note about role of items' identities in data providers (#1627)
vaadin/docs: articles/binding-data/data-provider.asciidoc

The easiest way to pass data to the listing component is to use an array or [classname]`List`.
You can create these easily yourself or pass values directly from your service layer.

> 
> 

Reviewer: ```suggestion
```

personGrid.setItems(dataProvider);
----

[[data-binding.data-provider.item-identifiers]]

> == Ensuring Items Identities are Stable and Unique

Reviewer 2: ```suggestion
== Ensuring That Item Identities Are Stable and Unique
```

When you bind items to a component, the identities of those items will be essential for the component to work properly.
For example, if you bind a list of `Person` objects to a `Grid`, the `Grid` will rely on the identities of the `Person` objects for various operations, such as for highlighting the selected rows and updating the data in an edited row.

**It is, therefore, important that the identities of the items are stable and unique.**

> To ensure that, your POJO class should implement [methodname]`hashCode()` and [methodname]`equals()` methods based on the invariant properties of the POJO.

Reviewer 3: Maybe "Bean class" instead of "POJO class"?

Reviewer: Is it actually a requirement for the items to be beans? Isn't it possible to use POJOs with custom getters and setters?

Reviewer 3: True, it's not strictly necessary. "Java class" might be more appropriate still.

Reviewer: Changed.

> To ensure that, your POJO class should implement [methodname]`hashCode()` and [methodname]`equals()` methods based on the invariant properties of the POJO.

Reviewer: ```suggestion
To ensure that, your Java class should implement [methodname]`hashCode()` and [methodname]`equals()` methods based on the invariant properties of the class.
```

> For example, the [classname]`Person` class in the example below have a `final long id` property that is used in the [methodname]`hashCode()` and [methodname]`equals()` methods as the identity of the [classname]`Person` object.

Reviewer 2: ```suggestion
For example, the [classname]`Person` class in the following example has a `final long id` property that is used in the [methodname]`hashCode()` and [methodname]`equals()` methods as the identity of the [classname]`Person` object.
```

This way even if the [classname]`Person` object has a property such as its `phoneNumber` modified, the identity of the [classname]`Person` object will not change, and a component like the `Grid` will still be able to identify the object correctly.

[source,java]
----
class Person {
    private final long id;
    private String name;
    private String phoneNumber;

    public Person(long id, String name, String phoneNumber) {
        this.id = id;
        this.name = name;
        this.phoneNumber = phoneNumber;
    }

    // Getters and setters omitted for brevity

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Person person = (Person) o;
        return id == person.id;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}
----

If you are using the Lombok library to generate [methodname]`hashCode()` and [methodname]`equals()` methods, you need to ensure that only the invariant properties are being used.

> You can do that by setting `onlyExplicitlyIncluded = true` for [annotationname]`@EqualsAndHashCode`, and then annotate the entity's `id` property with [annotationname]`@EqualsAndHashCode.Include`.

Reviewer 2: ```suggestion
You can do this by setting `onlyExplicitlyIncluded = true` for [annotationname]`@EqualsAndHashCode`, and then annotating the entity's `id` property with [annotationname]`@EqualsAndHashCode.Include`.
```

personGrid.setItems(dataProvider);
----

> == Beware of Lombok's @EqualsAndHashCode
> 
> By default, [classname]`@EqualsAndHashCode` generates [methodname]`equals()` and [methodname]`hashCode()` implementations using all of the entity's properties. 
> This behavior often interferes with Vaadin's data model.
> Vaadin expects the beans to be _invariant_, i.e. changing value of properties like `phoneNumber` should not affect the bean's identity.
> 
> There are two recommended ways to solve this problem:
> 
> . Set `onlyExplicitlyIncluded = true` for [classname]`@EqualsAndHashCode`, and then annotate the entity's `id` property with `@EqualsAndHashCode.Include`.
> +
> .SamplePerson.java
> [source,java]
> ----
> @Entity
> @EqualsAndHashCode(onlyExplicitlyIncluded = true)
> public class SamplePerson {
> 
>     @Id
>     @GeneratedValue
>     @Type(type = "uuid-char")
>     @EqualsAndHashCode.Include
>     private UUID id;
> 
>     private String phoneNumber;
> ----
> 
> . Extends [classname]`DataProvider` and overrides `getId()`
> +
> [source,java]
> ----
> class PersonDataProvider extends CallbackDataProvider<SamplePerson,Void> {
>     
>     ...
> 
>     @Override
>     public Object getId(SamplePerson item) {
>         return item.getId();
>     }
> }
> ----

Reviewer: ```suggestion
[[data-binding.data-provider.item-identifiers]]
== Ensuring Items Identities are Stable and Unique

When you bind items to a component, the identities of those items will be essential for the component to work properly.
For example, if you bind a list of `Person` objects to a `Grid`, the `Grid` will rely on the identities of the `Person` objects for various operations, such as for highlighting the selected rows and updating the data in an edited row.

**It is, therefore, important that the identities of the items are stable and unique.**
To ensure that, your POJO class should implement [methodname]`hashCode()` and [methodname]`equals()` methods based on the invariant properties of the POJO.
For example, the [classname]`Person` class in the example below have a `final long id` property that is used in the [methodname]`hashCode()` and [methodname]`equals()` methods as the identity of the [classname]`Person` object.
This way even if the [classname]`Person` object has a property such as its `phoneNumber` modified, the identity of the [classname]`Person` object will not change, and a component like the `Grid` will still be able to identify the object correctly.

[source,java]
----
class Person {
    private final long id;
    private String name;
    private String phoneNumber;

    public Person(long id, String name, String phoneNumber) {
        this.id = id;
        this.name = name;
        this.phoneNumber = phoneNumber;
    }

    // Getters and setters omitted for brevity

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Person person = (Person) o;
        return id == person.id;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}
----

If you are using the Lombok library to generate [methodname]`hashCode()` and [methodname]`equals()` methods, you need to ensure that only the invariant properties are being used.
You can do that by setting `onlyExplicitlyIncluded = true` for [annotationname]`@EqualsAndHashCode`, and then annotate the entity's `id` property with [annotationname]`@EqualsAndHashCode.Include`.

[source,java]
----
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
class Person {
    @EqualsAndHashCode.Include
    private final long id;
    private String name;
    private String phoneNumber;

    public Person(long id, String name, String phoneNumber) {
        this.id = id;
        this.name = name;
        this.phoneNumber = phoneNumber;
    }

    // Getters and setters omitted for brevity
}
----
```
